src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcherState.scala

```diff
@@ -88,6 +89,9 @@ case class BlockFetcherState(
         .ensure("Given headers are not sequence with already fetched ones")(_.head.number <= nextToLastBlock)
         .ensure("Given headers aren't better than already fetched ones")(_.last.number > lastBlock)
         .ensure("Given headers should form a sequence without gaps")(HeadersSeq.areChain)
+        .ensure("Given headers do not form a chain with already stored ones")(headers =>
+          (waitingHeaders.lastOption, headers.headOption).mapN(_ isParentOf _).getOrElse(true)
```

Comment:
Shouldn't we remove the conditions
- Given headers are not sequence with already fetched ones
- Given headers aren't better than already fetched ones
- Given headers do not form a chain with already stored ones

? The response of a peer isn't invalid if they don't fulfill this

If its a condition required by our waitingHeaders structure maybe we should guarantee this when appending them

Or at least we should stop blacklisting the peer if these validations fail, maybe we could have 2 separate errors, one for consistency of our structures and the other one for invalid responses? I'd separate it into two functions if possible

(note that the above overlaps with #771  )

Reply:
But these conditions are also to ensure consistency of internal state. Probably it makes more sense to change these errors into an ADT and make a decision about blacklisting basing on the result. I'm not sure if this PR is the place to do that change. WDYT?

Reply:
Given that we have blacklisting disabled I don't think it's critical to address this right away, we could do it after #771 gets merged (or even have this discussion there)

Reply:
I thought about it a bit (from different perspective though) and I came to conclusion that we need one of those (at least now):
  - build proper full branches on fetcher side
    That's because if mined block is accepted in other nodes - we'll receive it "properly" through network layer anyway and once we import branch got from fetcher - it'll be discarded (if lower/the same as the mined one) or used (if better blocks appeared on the network). This is essential from point of view of catching the network as it allows to have miner enabled and still be able to process blocks from network quite fast without big changes within code. That approach makes it a bit questionable to update `bestBlock` in fetcher state on receiving imported block, but it is still quite important when near top, otherwise node would fork too easy
  - clean fetcher state after receiving internal block (checkpoint or mined one) - while the easiest way to keep things in sync, etc. it would result in awful performance when catching the network (e.g. after turning node off for a night), because miner could effectively livelock/DoS the fetcher (especially in poor network conditions)

Given the above - I think that current approach present in code (building proper branches on fetcher side + disabling blacklisting due to useless responses) is the best until we extract component for managing tentative branches (where we could simply build multiple partial branches when near top, etc.)

Reply:
You mean then that we should close this PR? There's no way of reaching a better shape of it? I think there are some minor improvements that we can do without dropping the current approach

Either way, I'd definitely go for `build proper full branches on fetcher side` approach as the other one will lead to even more nasty workarounds. I think that maybe that maybe we could add the TentativeCompleteBranches only without removing the BlockQueue... but whatever we do about it will be on December

Reply:
`build proper full branches on fetcher side` - That's the approach implemented in this PR

Reply:
Oh I thought you meant doing something like the TBM with that

But then, wdyt about removing the conditions: `_.last.number > lastBlock`, `_.head.number <= nextToLastBlock`? Maybe I'm missing the point of having them

Reply:
i'm agree with Nico, i think the 2 first validations are no longer needed.